stratum/hal/lib/barefoot/bfrt_p4runtime_translator.h

```diff
@@ -0,0 +1,172 @@
+      LOCKS_EXCLUDED(lock_);
+  virtual ::util::Status PushForwardingPipelineConfig(
+      const ::p4::config::v1::P4Info& p4info) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::TableEntry> TranslateTableEntry(
+      const ::p4::v1::TableEntry& entry, bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::ActionProfileMember>
+  TranslateActionProfileMember(const ::p4::v1::ActionProfileMember& entry,
+                               bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::MeterEntry> TranslateMeterEntry(
+      const ::p4::v1::MeterEntry& entry, bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::DirectMeterEntry>
+  TranslateDirectMeterEntry(const ::p4::v1::DirectMeterEntry& entry,
+                            bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::CounterEntry> TranslateCounterEntry(
+      const ::p4::v1::CounterEntry& entry, bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::DirectCounterEntry>
+  TranslateDirectCounterEntry(const ::p4::v1::DirectCounterEntry& entry,
+                              bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::RegisterEntry> TranslateRegisterEntry(
+      const ::p4::v1::RegisterEntry& entry, bool to_sdk) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::PacketReplicationEngineEntry>
+  TranslatePacketReplicationEngineEntry(
+      const ::p4::v1::PacketReplicationEngineEntry& entry, bool to_sdk)
+      LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::PacketIn> TranslatePacketIn(
+      const ::p4::v1::PacketIn& packet_in) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::PacketOut> TranslatePacketOut(
+      const ::p4::v1::PacketOut& packet_out) LOCKS_EXCLUDED(lock_);
+  // A helper function which removes custom type from the P4Info.
+  // Which is useful for some components that requires the original spec from
+  // the P4 code.
+  // For example, the Packet-IO manager requires the real bitwidth information
+  // of controller header metadata.
+  virtual ::util::StatusOr<::p4::config::v1::P4Info> TranslateP4Info(
+      const ::p4::config::v1::P4Info& p4info);
+
+  static std::unique_ptr<BfrtP4RuntimeTranslator> CreateInstance(
+      BfSdeInterface* bf_sde_interface, int device_id,
+      bool translation_enabled) {
+    return absl::WrapUnique(new BfrtP4RuntimeTranslator(
+        bf_sde_interface, device_id, translation_enabled));
+  }
+
+ protected:
+  // Default constructor.
+  BfrtP4RuntimeTranslator()
+      : device_id_(0),
+        bf_sde_interface_(nullptr),
+        translation_enabled_(false),
+        pipeline_require_translation_(false) {}
+
+ private:
+  // Private constructor. Use CreateInstance() to create an instance of this
+  // class.
+  BfrtP4RuntimeTranslator(BfSdeInterface* bf_sde_interface, int device_id,
+                          bool translation_enabled)
```

Comment:
```suggestion
  BfrtP4RuntimeTranslator(bool translation_enabled, 
                          BfSdeInterface* bf_sde_interface, int device_id)
```

sde interface and device last.